datafiles/manager.py

```diff
@@ -34,22 +39,53 @@ def __init__(self, cls):
         self.model = cls
 
     def get(self, *args, **kwargs) -> Model:
-        fields = dataclasses.fields(self.model)
-        required = sum(1 for field in fields if isinstance(field.default, Missing))
-        missing_args = [Missing] * (required - len(args) - len(kwargs))
-        args = (*args, *missing_args)
-
         with hooks.disabled():
-            instance = self.model(*args, **kwargs)
+            instance = self.model.__new__(self.model)
+
+            # We **must** set up fields on the uninitialized instance which play a role
+            # in loading, e.g., those with placeholders in the pattern. Other fields
+            # which happen to have a value passed as an arg or kwarg are set as well, but
+            # these will eventually be loaded over anyhow.
+            fields = [field for field in dataclasses.fields(self.model) if field.init]
+            pattern = self.model.Meta.datafile_pattern
+            args_iter = iter(args)
+            for field in fields:
+                placeholder = f"{{self.{field.name}}}"
+
+                try:
+                    # we always need to consume an arg if it exists,
+                    # even if it's not one with a placeholder
+                    value = next(args_iter)
+                except StopIteration:
+                    value = kwargs.get(field.name, _NOT_PASSED)
+
+                if placeholder in pattern:
+                    if value is _NOT_PASSED:
+                        raise MissingPlaceholderArgumentError(
```

Comment:
I'l like to see a test added [here](https://github.com/jacebrowning/datafiles/blob/8f90d43e8d8dbf00a4faf34b9c72b739037961bb/datafiles/tests/test_manager.py#L58) that covers this line.

Rather than a custom exception, perhaps this should raise the same exception Python does when required arguments are missing?

Reply:
I thought to use a builtin exception, but the two commonly used relating to arguments are `TypeError` and `ValueError`. The docs have this to say:
> \[`TypeError` is\] raised when an operation or function is applied to an object of inappropriate type. The associated value is a string giving details about the type mismatch.
and
> Passing arguments of the wrong type (e.g. passing a [list](https://docs.python.org/3/library/stdtypes.html#list) when an [int](https://docs.python.org/3/library/functions.html#int) is expected) should result in a [TypeError](https://docs.python.org/3/library/exceptions.html#TypeError), but passing arguments with the wrong value (e.g. a number outside expected boundaries) should result in a [ValueError](https://docs.python.org/3/library/exceptions.html#ValueError).

In this case, no argument is passed, how could it be an inappropriate type?

I suppose the CPython implementation itself goes against this though, considering the following example:
```py
>>> def test(a, b, *, c):
...   print(a,b,c)
...
>>> test(1,2,3)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: test() takes 2 positional arguments but 3 were given
>>> test(1,2, c=3)
1 2 3
>>> test(1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: test() missing 1 required positional argument: 'b'
>>> test(1,2)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: test() missing 1 required keyword-only argument: 'c'
```

I suppose `TypeError` it is.

Reply:
Can I put `describe_get` up a bit, [here](https://github.com/jacebrowning/datafiles/blob/8f90d43e8d8dbf00a4faf34b9c72b739037961bb/datafiles/tests/test_manager.py#L39)?

Reply:
Yeah, that works!

Reply:
What is `expect`? I can't find any info about it being built into `pytest`?

Reply:
Ah, I expect it's `pytest-expecter`.

Reply:
https://pytest-expecter.readthedocs.io/

Reply:
Variants of `poetry run pytest datafiles/tests/test_manager.py` aren't working for me. Any tips on getting just this file to run?

Reply:
That command _should_ work. Does `poetry run pytest -k manager` work?

Reply:
That works.